### PR TITLE
#139, #145, #147: Database schema, model logic and scheduled jobs

### PR DESCRIPTION
**Pull request by DanBondarenko**
Overview:
* Updated the database schema to support archiving and future deactivation. I deleted the existing `status` column no longer made sense, since the status is now computed from the values of other columns.
* Created methods for working with statuses in the `InstitutionUser` class.
* Created a scheduled nightly job, which deleted roles of users who have become deactivate.

This work is part of the following package of tasks:
* https://github.com/keeleinstituut/tv-tolkevarav/issues/139
* https://github.com/keeleinstituut/tv-tolkevarav/issues/145
* https://github.com/keeleinstituut/tv-tolkevarav/issues/147

### CODE REVIEW

File: application/database/migrations/2023_05_15_121457_institution_user_columns_for_deactivation_activation_archival.php
Comment (kadmit) [inline application/database/migrations/2023_05_15_121457_institution_user_columns_for_deactivation_activation_archival.php:1]: **[blocking]** @DanBondarenko I would recommend adding indexes for the `institution_users` table. Would be good to have an index that will include `archived_at` and `deactivation_date` as these fields are included in the query for fetching active users.

**Review by kadmit — CHANGES_REQUESTED** [review]

File: application/database/migrations/2023_05_15_121457_institution_user_columns_for_deactivation_activation_archival.php
Comment (DanBondarenko) [reply application/database/migrations/2023_05_15_121457_institution_user_columns_for_deactivation_activation_archival.php:1]: @kadmit That is a great idea! Thank you for the suggestion. I added it in the latest revision.

**Comment by DanBondarenko** [comment]
@kadmit Just notifying that I implemented your suggestion, so you may re-review whenever you get time. :)

File: application/app/Console/Kernel.php [hunk 1]
```diff
@@ -12,7 +13,11 @@ class Kernel extends ConsoleKernel
      */
     protected function schedule(Schedule $schedule): void
     {
-        // $schedule->command('inspire')->hourly();
+        $schedule
+            ->command(DetachRolesFromDeactivatedUsers::class)
+            ->timezone('Europe/Tallinn')
+            ->onOneServer()
+            ->daily();
```
[the remark this answers is not in the document]
Comment (DanBondarenko) [reply application/app/Console/Kernel.php:21]: The `DetachRolesFromDeactivatedUsers` job behavior is not affected by server timezone. It compares the current date in Estonia to the date in database  (e.g. `2020-12-12`).

File: application/app/Models/InstitutionUser.php [hunk 2]
```diff
@@ -97,4 +104,71 @@ public function roles(): BelongsToMany
...
+            return InstitutionUserStatus::Deactivated;
+        }
+
+        return InstitutionUserStatus::Active;
+    }
+
+    public function isDeactivated(): bool
+    {
+        return filled($this->deactivation_date)
+            && ! $this->deactivation_date->isAfter(DateUtil::currentEstonianDateAtMidnight());
```
[the remark this answers is not in the document]
Comment (DanBondarenko) [reply application/app/Models/InstitutionUser.php:146]: Could you specify what do you mean by hardcoding the timezone? The timezone is not saved anywhere AFAIK – the database only has a date-only value (e.g. `2020-12-12`).

File: application/app/Models/Scopes/ExcludeDeactivatedInstitutionUsersScope.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+use Illuminate\Support\Facades\Date;
+
+class ExcludeDeactivatedInstitutionUsersScope implements Scope
+{
+    public function apply(Builder $builder, Model $model): void
+    {
+        $builder->where(
+            fn ($groupedClause) => $groupedClause
+                ->whereNull('deactivation_date')
+                ->orWhereDate('deactivation_date', '>', Date::now('Europe/Tallinn'))
```
[the remark this answers is not in the document]
Comment (DanBondarenko) [reply application/app/Models/Scopes/ExcludeDeactivatedInstitutionUsersScope.php]: Well spotted!

**Review by kadmit — APPROVED** [review]
LGTM

**Comment by DanBondarenko** [comment]
For future reference, I was requested to make to remove the methods of the `DateUtil` class as there was a concern that it might invite future developers to overzealously specify the Estonian timezone everywhere, which would not be good practice. Therefore I inlined the methods, resulting in a bit of duplication.

**Comment by DanBondarenko** [comment]
@thenouan Ready for your re-review

**Review by thenouan — APPROVED** [review]

**Review by thenouan — APPROVED** [review]

**Pull request merged**